### PR TITLE
*: make IfWatcher::new synchronous

### PR DESCRIPTION
**Pull request by elenaf9**
Make `IfWatcher::new` synchronous. The only reason why it is currently async is because the `linux` implementation executed a [`AddressGetRequest`](https://docs.rs/rtnetlink/latest/rtnetlink/struct.AddressGetRequest.html#method.execute) in `new` to obtain the initial list of ip addresses. By chaining this stream together with the existing message stream we can make `linux::IfWatcher::new` sync and instead move the polling into the `Stream` impl.
Motivation for this change is that `IfWatcher::new` being async currently causes some complexity in rust-libp2p. Concretely, it forces to wrap the `IfWatcher` in an enum [1] and first poll the future returned by `IfWatcher::new` [2] before we can poll the `IfWatcher` itself.
- [1] https://github.com/libp2p/rust-libp2p/blob/3da8b423c282383b7c55b2eda7050bd991a2813b/transports/tcp/src/lib.rs#L608-L611
- [2] https://github.com/libp2p/rust-libp2p/blob/3da8b423c282383b7c55b2eda7050bd991a2813b/transports/tcp/src/lib.rs#L747-L764

### CODE REVIEW

**Review by mxinden — APPROVED** [review]
:+1: Cool. This looks good to me.

**Comment by mxinden** [comment]
@elenaf9 could you resolve the merge conflicts?

**Review by mxinden — CHANGES_REQUESTED** [review]
Oh, missing changelog entry.

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,8 +8,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 - Add `IfWatcher::poll_next`. Implement `Stream` instead of `Future` for `IfWatcher`. See [PR 23].
+- Make `IfWatcher::new` synchronous. See [PR 24].
```
Comment (elenaf9) [inline CHANGELOG.md:11]: Not sure how to describe this better.

Comment (mxinden) [reply CHANGELOG.md:11]: Looks good to me.

**Review by mxinden — APPROVED** [review]

**Comment by mxinden** [comment]
@elenaf9 should I cut a release already, or do you expect other changes in the near future?

**Comment by elenaf9** [comment]
> @elenaf9 should I cut a release already, or do you expect other changes in the near future?

I am still looking into simplifying the `IfWatcher` integration in rust-libp2p. There may still be some minor changes coming. I will ping you once I am through and we can cut a release. Thanks!

**Pull request merged**